ext/liquid_c/block.c

```diff
@@ -79,28 +68,45 @@ static VALUE rb_block_parse(VALUE self, VALUE tokens, VALUE options)
                 rb_ary_push(nodelist, str);
 
                 if (rb_ivar_get(self, intern_blank) == Qtrue) {
-                    const char *end = token.str + token.length;
+                    const char *end = token.str_full + token.len_full;
 
-                    if (read_while(token.str, end, rb_isspace) < end)
+                    if (read_while(token.str_full, end, rb_isspace) < end)
                         rb_ivar_set(self, intern_blank, Qfalse);
                 }
                 break;
             }
             case TOKEN_VARIABLE:
             {
-                VALUE args[2] = {rb_enc_str_new(token.str + 2 + token.lstrip, token.length - 4 - token.lstrip - token.rstrip, utf8_encoding), options};
+                VALUE args[2] = {rb_enc_str_new(token.str_trimmed, token.len_trimmed, utf8_encoding), parse_context};
                 VALUE var = rb_class_new_instance(2, args, cLiquidVariable);
                 rb_ary_push(nodelist, var);
                 rb_ivar_set(self, intern_blank, Qfalse);
                 break;
             }
             case TOKEN_TAG:
             {
-                const char *start = token.str + 2 + token.lstrip, *end = token.str + token.length - 2 - token.rstrip;
+                const char *start = token.str_trimmed, *end = token.str_trimmed + token.len_trimmed;
 
                 // Imitate \s*(\w+)\s*(.*)? regex
                 const char *name_start = read_while(start, end, rb_isspace);
                 const char *name_end = read_while(name_start, end, is_id);
+                long name_len = name_end - name_start;
+
+                if (name_len == 0) {
+                    VALUE str = rb_enc_str_new(token.str_trimmed, token.len_trimmed, utf8_encoding);
+                    return rb_yield_values(2, str, str);
+                }
+
+                if (strncmp(name_start, "liquid", name_len < 6 ? name_len : 6) == 0) {
```

Comment:
Currently this will match truncations of `"liquid"` like `"l"` or `"liqui"`.  It should be strictly equal like in the ruby implementation

```suggestion
                if (name_len == 6 && strncmp(name_start, "liquid", 6) == 0) {
```